File: clients/android/app/src/main/java/tv/plurx/app/ui/LibraryMerge.kt
package tv.plurx.app.ui

import tv.plurx.app.data.Item

/** Server order across paged, individually sorted library cursors. */
internal class LibraryMerge(ids: List<Long>, private val sort: String) {
    private data class Cursor(
        val id: Long,
        var offset: Int = 0,
        var total: Int = 0,
        val buffer: ArrayDeque<Item> = ArrayDeque(),
        var exhausted: Boolean = false,
    )

    private val cursors = ids.map(::Cursor)
    val decided = mutableListOf<Item>()
    var missingSortKey = false
        private set
    val loadedCount get() = cursors.sumOf { it.offset }
    val total get() = cursors.sumOf { it.total }
    val complete get() = cursors.all { it.exhausted && it.buffer.isEmpty() }
    val nextRequest: Pair<Long, Int>?
        get() = cursors.firstOrNull { !it.exhausted && it.buffer.isEmpty() }?.let { it.id to it.offset }

    fun receive(id: Long, items: List<Item>, total: Int, limit: Int = 200) {
        val cursor = cursors.first { it.id == id }
        missingSortKey = missingSortKey || items.any { it.sort_title == null }
        cursor.buffer.addAll(items)
        cursor.offset += items.size
        cursor.total = total
        cursor.exhausted = items.isEmpty() || cursor.offset >= total || items.size < limit
        drainDecidable()
    }

    private fun drainDecidable() {
        while (nextRequest == null) {
            val heads = cursors.filter { it.buffer.isNotEmpty() }
            if (heads.isEmpty()) return
            val selected = heads.minWithOrNull { a, b -> compare(a.buffer.first(), b.buffer.first(), sort) }!!
            decided += selected.buffer.removeFirst()
        }
    }

    companion object {
        fun compare(a: Item, b: Item, sort: String): Int {
            fun title(): Int = compareUtf8(a.sort_title.orEmpty(), b.sort_title.orEmpty())
            fun <T : Comparable<T>> descending(left: T?, right: T?): Int =
                when {
                    left == right -> 0
                    left == null -> 1
                    right == null -> -1
                    else -> right.compareTo(left)
                }
            val primary = when (sort) {
                "added" -> descending(a.added_at, b.added_at)
                "year" -> descending(a.year, b.year).takeIf { it != 0 } ?: title()
                "resolution" -> descending(a.resolution ?: -1, b.resolution ?: -1).takeIf { it != 0 } ?: title()
                "recorded" -> descending(a.recorded_at, b.recorded_at).takeIf { it != 0 } ?: title()
                else -> title()
            }
            if (primary != 0) return primary
            return if (sort == "added") b.id.compareTo(a.id) else a.id.compareTo(b.id)
        }

        private fun compareUtf8(a: String, b: String): Int {
            val left = a.toByteArray(Charsets.UTF_8)
            val right = b.toByteArray(Charsets.UTF_8)
            for (i in 0 until minOf(left.size, right.size)) {
                val comparison = (left[i].toInt() and 0xff).compareTo(right[i].toInt() and 0xff)
                if (comparison != 0) return comparison
            }
            return left.size.compareTo(right.size)
        }
    }
}

File: clients/apple/Sources/LibraryMerge.swift
import Foundation

/// The server's total order across individually paged libraries. A buffered
/// head from every unfinished cursor is required before any item is decided.
struct LibraryMerge {
    struct Cursor {
        let libraryId: Int
        var offset = 0
        var total = 0
        var buffer: [Item] = []
        var exhausted = false
    }

    let sort: LibrarySort
    private(set) var cursors: [Cursor]
    private(set) var decided: [Item] = []
    private(set) var missingSortKey = false

    init(libraryIds: [Int], sort: LibrarySort) {
        self.sort = sort
        cursors = libraryIds.map { Cursor(libraryId: $0) }
    }

    var loadedCount: Int { cursors.reduce(0) { $0 + $1.offset } }
    var total: Int { cursors.reduce(0) { $0 + $1.total } }
    var complete: Bool { cursors.allSatisfy { $0.exhausted && $0.buffer.isEmpty } }

    var nextRequest: (libraryId: Int, offset: Int)? {
        guard let cursor = cursors.first(where: { !$0.exhausted && $0.buffer.isEmpty }) else { return nil }
        return (cursor.libraryId, cursor.offset)
    }

    mutating func receive(libraryId: Int, items: [Item], total: Int, limit: Int = 200) {
        guard let index = cursors.firstIndex(where: { $0.libraryId == libraryId }) else { return }
        missingSortKey = missingSortKey || items.contains { $0.sortTitle == nil }
        cursors[index].buffer.append(contentsOf: items)
        cursors[index].offset += items.count
        cursors[index].total = total
        cursors[index].exhausted = items.isEmpty || cursors[index].offset >= total || items.count < limit
        drainDecidable()
    }

    mutating func drainDecidable() {
        while nextRequest == nil {
            let heads = cursors.indices.filter { !cursors[$0].buffer.isEmpty }
            guard let first = heads.first else { return }
            let selected = heads.dropFirst().reduce(first) { best, candidate in
                Self.precedes(cursors[candidate].buffer[0], cursors[best].buffer[0], sort: sort) ? candidate : best
            }
            decided.append(cursors[selected].buffer.removeFirst())
        }
    }

    static func precedes(_ lhs: Item, _ rhs: Item, sort: LibrarySort) -> Bool {
        func title() -> Bool? {
            let a = Array((lhs.sortTitle ?? "").utf8)
            let b = Array((rhs.sortTitle ?? "").utf8)
            if a == b { return nil }
            return a.lexicographicallyPrecedes(b)
        }
        func descending<T: Comparable>(_ a: T?, _ b: T?) -> Bool? {
            if a == b { return nil }
            guard let a else { return false }
            guard let b else { return true }
            return a > b
        }
        switch sort {
        case .title:
            if let order = title() { return order }
        case .added:
            if let order = descending(lhs.addedAt, rhs.addedAt) { return order }
            return lhs.id > rhs.id
        case .year:
            if let order = descending(lhs.year, rhs.year) { return order }
            if let order = title() { return order }
        case .resolution:
            if let order = descending(lhs.resolution ?? -1, rhs.resolution ?? -1) { return order }
            if let order = title() { return order }
        case .recorded:
            if let order = descending(lhs.recordedAt, rhs.recordedAt) { return order }
            if let order = title() { return order }
        }
        return lhs.id < rhs.id
    }
}
